pkg/syncer/upsync/upsync_process.go

```diff
@@ -0,0 +1,183 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	ResourceVersionAnnotation = "kcp.dev/resource-version"
+)
+
+func (c *Controller) pruneUpstreamResource(ctx context.Context, gvr schema.GroupVersionResource, key string) error {
+	logger := klog.FromContext(ctx)
+	upstreamNamespace, upstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+	err = c.upstreamClient.Cluster(c.syncTargetWorkspace).Resource(gvr).Namespace(upstreamNamespace).Delete(ctx, upstreamName, metav1.DeleteOptions{})
+	if err != nil {
+		logger.Error(err, "Could not remove resource upstream")
+		return nil
+	}
+	return nil
+}
+
+func (c *Controller) processDownstreamResource(ctx context.Context, gvr schema.GroupVersionResource, key string) error {
+	logger := klog.FromContext(ctx)
+	downstreamNamespace, downstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+
+	logger = logger.WithValues(DownstreamNamespace, downstreamNamespace, DownstreamName, downstreamName)
+
+	var upstreamLocator *shared.NamespaceLocator
+	var locatorExists bool
+	var downstreamObject interface{}
+	// Namespace scoped resource
+	if downstreamNamespace != "" {
+		var nsObj runtime.Object
+		if nsObj, err = c.downstreamNamespaceLister.Get(downstreamNamespace); err != nil {
+			logger.Error(err, "Error getting downstream Namespace from downstream namespace lister", "ns", downstreamNamespace)
+		}
+		nsMeta, ok := nsObj.(metav1.Object)
+		if !ok {
+			logger.Info(fmt.Sprintf("Error: downstream ns expected to be metav1.Object got %T", nsObj))
+		}
+
+		upstreamLocator, locatorExists, err = shared.LocatorFromAnnotations(nsMeta.GetAnnotations())
+		if err != nil {
+			logger.Error(err, "Error ")
+		}
+
+		if !locatorExists || upstreamLocator == nil {
+			return nil
+		}
+		downstreamObject, err = c.downstreamClient.Resource(gvr).Namespace(downstreamNamespace).Get(ctx, downstreamName, metav1.GetOptions{})
+		if err != nil {
+			logger.Error(err, "Could not find the resource downstream")
+		}
+	}
+	if downstreamNamespace == "" {
+		downstreamObject, err = c.downstreamClient.Resource(gvr).Get(ctx, downstreamName, metav1.GetOptions{})
+		if err != nil {
+			logger.Error(err, "Could not find the resource downstream")
+		}
+		objMeta, ok := downstreamObject.(metav1.Object)
+		if !ok {
+			logger.Info("Error: downstream cluster-wide res not of metav1 type")
+		}
+		if upstreamLocator, locatorExists, err = shared.LocatorFromAnnotations(objMeta.GetAnnotations()); err != nil {
+			logger.Error(err, "Error decoding annotations on DS cluster-wide res")
+		}
+
+		if !locatorExists || upstreamLocator == nil {
+			return nil
+		}
+	}
+
+	u, ok := downstreamObject.(*unstructured.Unstructured)
```

Comment:
If we know it will be an `unstructured` object, let's convert it at the beginning of the function. So you won't have to convert to `metav1`

Reply:
updated!